tracpro/contacts/fields.py

```diff
@@ -41,14 +57,110 @@ def render(self, name, value, attrs=None):
 
 
+def urn_exists_in_rapidpro(client, uuid, urn):
+    """
+    Return True if a remote contact with this urn
+    exists in RapidPro and has a different uuid than uuid.
+    """
+
+    contacts = client.get_contacts(urn=urn)
+    return contacts and contacts[0].uuid != uuid
+
+
+def urn_already_used(client, uuid, urn):
+    """Return True if a local or remote contact
+        with this urn exists and has a different uuid than uuid"""
+
+    from .models import Contact
+    # RapidPro checks for the numeric id without handle
+    # we are going to mimic that here locally
+    if urn.startswith('twitterid'):
+        contacts = Contact.objects.filter(urn__startswith=urn.split('#')[0])
+        if uuid:
+            contacts = contacts.exclude(uuid=uuid)
+    else:
+        contacts = Contact.objects.filter(urn=urn)
+        if uuid:
+            contacts = contacts.exclude(uuid=uuid)
+
+    return contacts.exists() or urn_exists_in_rapidpro(client, uuid, urn)
+
+
+def validate_twitter_handle(client, uuid, handle, from_id=False):
+    handle_regex = r'^[a-zA-Z0-9_]{1,15}$'
+    if not re.match(handle_regex, handle):
+        raise ValidationError(
+            _("This is not a valid Twitter handle.  A valid handle has only "
+              "letters A-Z, numbers 0-9, and underscores (_), and has up to 15 "
+              "characters."))
+    elif not from_id and urn_already_used(client, uuid, 'twitter:%s' % handle):
+        raise ValidationError(
+            _("This Twitter Handle is already being used by another contact."))
+
+
+def validate_numeric_twitter_id(numeric_id):
+    if not numeric_id.isdigit():  # True only if all chars are digits
+        raise ValidationError(
+            _("This is not a valid numeric ID.  A valid numeric ID has only "
+              "numbers 0-9."))
+
+
+def validate_twitter_id(client, uuid, twitter_id):
+    if '#' in twitter_id:
+        numeric_id, handle = twitter_id.split('#', 1)
+        validate_twitter_handle(client, uuid, handle, True)
+        validate_numeric_twitter_id(numeric_id)
+        if urn_already_used(client, uuid, 'twitterid:%s' % twitter_id):
+            raise ValidationError(
+                _("This Twitter ID is already being used by another contact."))
+    else:
+        raise ValidationError(
+            _("This is not a valid numeric ID.  The correct format is: "
+              "<numeric id>#<Twitter handle>"))
+
+
+def validate_phone_used(client, uuid, parsed):
+    phone_number = '%s%s%s' % ('+', parsed.country_code, parsed.national_number)
+    urn = 'tel:%s' % phone_number
+
+    if urn_already_used(client, uuid, urn):
+        raise ValidationError(
+             _("This phone number is already being used by another contact."))
+    else:
+        return urn
+
+
+def validate_phone(client, uuid, number):
+    try:
+        parsed = parse(number)
+    except NumberParseException as error:
+        raise ValidationError(
+            _(PHONE_PARSE_ERROR[error.error_type]))
```

Comment:
Sorry I just noticed this: the ugettext_lazy() call (here aliased to ``_``) needs to be wrapped around the literal text that needs translating. It won't work wrapped around a variable reference or more complicated expression.

(I can explain this IRL if you want, it's too much to type here :-))